### PR TITLE
revert #168, fix #167

### LINKED ISSUE #167: Cannot focus on input element inside iframe when the editor has focus.
## Describe the problem

Cannot focus on `input` element inside `iframe` when the editor has focus.

- Reproduction
  1. visit https://learn.svelte.dev/tutorial/named-form-actions
  2. click the editor
  3. click the input element in iframe

https://user-images.githubusercontent.com/29677552/209102192-762c2cde-b64d-4920-bd7f-d58d02884585.mov

This is probably related to #161. 
I suppose the intent of #161 is to consider `pointerdown` event as a human operation. However, events that occur inside `iframe` cannot be handled from outside `iframe`.

https://github.com/sveltejs/learn.svelte.dev/blob/8d0376cec2f0a6117726d4af2b5cb88c115353ab/src/routes/tutorial/%5Bslug%5D/Editor.svelte#L243-L249

https://github.com/sveltejs/learn.svelte.dev/blob/8d0376cec2f0a6117726d4af2b5cb88c115353ab/src/routes/tutorial/%5Bslug%5D/Editor.svelte#L266-L272

## Describe the proposed solution

when `body` element inside `iframe` gets focus, set focus on the editor. (using `postMessage`)

https://github.com/tomoam/learn.svelte.dev/blob/e291cb8c0d08f011d358427f7990fe7f2e6257d3/content/tutorial/common/src/__client.js#L93-L106

https://github.com/tomoam/learn.svelte.dev/blob/e291cb8c0d08f011d358427f7990fe7f2e6257d3/src/routes/tutorial/%5Bslug%5D/Editor.svelte#L236-L242

### PR DESCRIPTION
**Pull request by Rich-Harris**
#168 introduced a bug with focus management — if you click on a link inside the iframe, it focuses the editor. This reverts that PR and fixes #167 by emitting a message when the iframe receives a `pointerdown` event

### CODE REVIEW

File: src/routes/tutorial/[slug]/state.js [hunk 1]
```diff
@@ -1,5 +1,17 @@
 import { derived, writable } from 'svelte/store';
 
+/**
+ * The iframe sometimes takes focus control in ways we can't prevent
+ * while the editor is focused. Refocus the editor in these cases.
+ * This boolean tracks whether or not the editor should be refocused.
+ */
+export let preserve_editor_focus = false;
```
Comment (dummdidumm) [inline src/routes/tutorial/[slug]/state.js:8]: This state is only relevant to the editor, I'd suggest keeping it in there. We already have `svelte:window` in `Editor.svelte`, which we can use to listen to messages as well.

**Review by dummdidumm — COMMENTED** [review]
Looks good, only got one suggestion about code organization 👍

File: src/routes/tutorial/[slug]/state.js (hunk 1 above)
Comment (Rich-Harris) [reply src/routes/tutorial/[slug]/state.js]: good catch, updated

**Review by dummdidumm — APPROVED** [review]

**Comment by tomoam** [comment]
Thank you, but it appears that the issue similar to #167 has occurred.
I reopen #167.

https://user-images.githubusercontent.com/29677552/214755592-dcac0d23-84f9-4da1-a90e-f4d205ec8524.mov

**Comment by dummdidumm** [comment]
Thanks for reporting, I pushed a fix that hopefully resolved this

**Comment by tomoam** [comment]
@dummdidumm 
Thank you very much. However, it now appears that another issue has occurred.
When the iframe reloads while I am typing, the text I type is not entered correctly.


https://user-images.githubusercontent.com/29677552/214850880-cb09cf60-ab7d-4f74-ade7-a6e2bf25bbd9.mov

I typed `export let data;`, but entered `export let dat;`. The `a` is missing.
Perhaps text entered during the 100ms (the time of refocusing editor) would be lost.
I'll try to figure out how to fix it too.

**Comment by dummdidumm** [comment]
God this iframe stuff is so annoying - thanks for taking another look. Maybe also write down a list of manual tests for future reference so we know that further tweaks don't introduce regressions.

**Comment by tomoam** [comment]
@dummdidumm 
I have created tests with Playwright that cover #150, #167, and this issue ([tomoam/learn.svelte.dev:add-test](https://github.com/sveltejs/learn.svelte.dev/compare/main...tomoam:learn.svelte.dev:add-test)).
Based on this tests, I will create a PR to fix this issue. If you don't need the PR, let me know.

**Pull request merged**